Implement docker update command

hqhq opened the pull request:
fixes #6323

It's used for change properties of one or more containers, we only
support resource configs for now. It can be extended in the future.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

---

The original PR #11442, compared to the former one, changelog:
- support set for both running and stopped containers
- support more cgroup configs as the new ones were merged
- move some container config change logic to container daemon/container_xxx.go
- addressed all issues pointed by @duglin 

We have strong needs for this feature for a long time, see #3285. It's really useful feature and necessary for most product usage.

@icecrime @crosbymichael as we discussed on DockerCon, we should reconsider this feature, I have pinged you guys several times hope we can reopen the original one but got no reply, so I opened this new PR, hope it's ok. If the old one can be reopened, I'm ok to close this and continue the original discussion.

calavera commented:
So, my take on this new command is that it doesn't really address any of the issues described there, per-se:

```
Unfortunately, docker doesn't currently seem to support this. Since you can't re-run a stopped container with docker run, you would need to commit the container to an image and then re-run it.

Unfortunately, docker commit clears the container's configuration, so you need to explicitly set a startup configuration using the -run commandline argument, unless you don't mind manually specifying the entrypoint/cmd et cetera each time you run this new image with docker run (which I don't think is very user friendly). As a workaround, you can inspect the container, copy the Config, and paste the config into the -run parameter of docker commit, but this isn't really user-friendly either.
```

All those issues will still be present if we add this new command, but people will have to learn how to use `docker set` to work around them.

From my point of view, the main pain points are caused by bad usability in other commands. I'd personally prefer to have concrete solutions to those issues before adding a new command, but I can be convinced that we need a new command if there is no other way to solve those issues.

tiborvass commented:
I have a weak +1 for dynamically changing resource configs on both running and stopped containers, however I am not exactly sure if we would want (or even able to) change other settings (e.g. environment variables). Therefore, I don't know if `set` is the right keyword. Maybe it is.

hqhq commented:
@calavera 

> Unfortunately, docker doesn't currently seem to support this. Since you can't re-run a stopped container with docker run, you would need to commit the container to an image and then re-run it.
We don't need to commit the container to an image and re-run it, we just do:
```
$ docker stop xxx
$ docker set xxx
$ docker start xxx
```

> Unfortunately, docker commit clears the container's configuration, so you need to explicitly set a startup configuration using the -run commandline argument, unless you don't mind manually specifying the entrypoint/cmd et cetera each time you run this new image with docker run (which I don't think is very user friendly). As a workaround, you can inspect the container, copy the Config, and paste the config into the -run parameter of docker commit, but this isn't really user-friendly either.
As above, this is not the common use case, for people who want to dynamically change properties, they don't want to delete or commit current container and re-run it.

hqhq commented:
@tiborvass I'm open with the command name :smile:

selvik commented:
@hqhq I am trying to understand you related prior PR here: https://github.com/docker/docker/pull/11442 and the feedback received.

Do you agree with the suggestion that using --cgroup-parent enables us to change resource configs of a running container? I don't yet see how this could be done.

Since only docker run accepts the --cgroup-parent flag, the only way I can actually change the config. of a container that is already started is to directly write to the cgroup filesystem on the docker host.

Would you have an example of how --cgroup-parent can be used to change config of a running container? I have asked the same question in that PR too, but wasn't sure if a closed PR intimates those involved.

Thanks for any hints!

calavera commented:
> We don't need to commit the container to an image and re-run it, we just do:

but if we allow to run stopped containers with `docker run` we would not need to add a new command, right?

hqhq commented:
@selvik You can always change resource configs of running containers by changing their cgroup files, that has nothing to do with Docker, but usually the problem is it relayed on users understand cgroup details, know the cgroup path of their containers, and know which files to be changed, `--parent-cgroup` option just make user know what path is the container in.

So I didn't mean `--cgroup-parent` can resolve the problem, it just did very litter help, so we still need a more friendly docker command so users don't need to know there cgroup details.

hqhq commented:
@calavera Sort of. But it only resolve the problem that change configs for stopped containers, not for running containers, and I think that'll make it more complicated, because not all options of `docker run` can be changed. So I don't think that's a good idea to make `docker run` can start a stopped container.

selvik commented:
@hqhq I agree, we're on the same page then. docker set would be very useful for us. Looking forward to this PR being merged! Thanks for working on it!

cnaize commented:
> @hqhq 
> We don't need to commit the container to an image and re-run it, we just do:
> $ docker stop xxx
> $ docker set xxx
> $ docker start xxx
Is it mean that I have to `stop` container if I want to change `memory` limit, for example?

I need to change limits on `running` container, how to do it? Only by `cgroup`?
Could you please provide an example?

If I can help you to implement changing `limits` on the fly, let me know.

hqhq commented:
@cnaize No you don't need to stop container, you can just use `docker set -m 300M my_container` if you want to change memory limit for a running container.

My approach support changing configs for both running and stopped containers. You don't need to know if a container is running or stopped, for stopped container, the changed configs will be effective when it started again.

Currently, the `configs` I mentioned only mean resource configs, but it can be expended in the future.

vdemeester commented:
How are we on that ? 

- I think using flags (e.g. `docker set --cpu-period … …`) is not that good ; and I would tend to prefer something like `docker set cpu-period … …` but it really is a nit I guess. It would allow to not have to declare all the thing that can be change into the `client` code, but on the other hand it would probably expose too much stuff.
- The command name `set` feels way too *generic* I think, it should probably be `set-config` or something like that. WDYT ?

@hqhq needs a rebase :sweat: 

/cc @cpuguy83 @icecrime @runcom

tiborvass commented:
*Collective review*

There are valid usecases detailed below. The `docker set` solution would be a BIG change that would need @shykes's input before we can move on.

For the sake of the discussion, we added more questions to explore different solutions for the usecases.

Some valid usecases:
- change ports dynamically:
  * I start developing in a container, and later I realize I want to expose a port.
  * add a debug port and then remove it
  * could `docker network` address this usecase? `docker network` handles networks and not ports. Is that an issue?
- change cgroups:
  * one solution for this is to link from documentation to external tools that can change actual cgroups for docker containers
- change volumes/label dynamically?
  * **We need an answer to the following question: when is it appropriate to change a container property vs requiring to spawn a new container with different properties?**
  * for labels specifically, should there be a label-specific UI if we were to also include changing daemon labels and not just container labels?

One solution would be to add a new `docker set` command but restrict it to whatever properties we agree on.

tiborvass commented:
Forgot to mention environment variables: I'm not even sure how you can do that without restarting the process at which point you might as well restart the container.

duglin commented:
yup - some config changes may not take effect until there's a restart of the container, but I can see that for some scenarios that's better than having to start from scratch with a new container.

hqhq commented:
Yes `docker set` has to be restricted to whatever properties we agreed on, not all properties `docker run` supported can be changed dynamically. And some of them need a restart to make them effect, I think we can give users some messages for these configs when they using them for a running container.
@vdemeester Seems this PR is still in design review stage, I'd like to rebase when it's in code review stage, is that OK?

rutsky commented:
Where should I address request to implement `docker set` for specific configuration options?
E.g. I would like to have ability to change `RestartPolicy`, and change of this property shouldn't break anything (in comparison with for example changing environment variables).

wdx900 commented:
Hi，我现在希望在container running的情况下去修改容器的端口映射，我看到docker set可以实现CPU和MEM的动态修改，但是如何去动态修改他的端口映射，你有什么建议吗，谢谢

hqhq commented:
@rutsky If we can get this PR merged, your request would be easy to implement. You can open new issue for that or post here.

@wdx900 It not supported yet, `docker set` command is still under discussion, change ports dynamically is doable with `docker set`, let's hope this PR get permission soon, so we can move on. And please, post English only here :)

xenda commented:
Is there by any chances any update on this discussion? This is something that would really help us solve several issues. I'm seeing it's failing in the windows build, so maybe that's something additional to be taken care of.

hqhq commented:
ping @shykes @crosbymichael since we want your input before move on.

thaJeztah commented:
Trying to move this in a direction, we discussed "mutability" of containers / container-properties in last Thursday's maintainers meeting. @duglin @vdemeester @jfrazelle and I are going to work on this, and (try to) get some input from @shykes.

duglin commented:
I believe this issue was discussed during the maintainers summit at dockercon eu - perhaps one of the folks at that table could summarize the results of those talks?  @tiborvass @runcom @vdemeester   I think you guys were at that table.

vdemeester commented:
@duglin Yes, will do shortly (just have to take some times to summarize the notes) :wink:

vdemeester commented:
To summarize : The main discussion we had (with @thaJeztah @tiborvass @unclejack @runcom @coolljt0725) was about what should or should not be mutable on the container(s). As of now, there is only one mutable element of the container, it is the network(s) — a running container can connect/disconnect networks.

We decided to allow to set what we called `resources`, which consists of cgroup *thingies* for now, hence the following PR #18073. The only *allowed* mutable elements of a container are in `HostConfig` and precisely in `Resources` (see the [struct](https://github.com/docker/docker/blob/master/runconfig/hostconfig.go#L169)).

On the name of the command, we didn't really found a better name than `set`. It stays a *root* command (no `container set` subcommand for ex.) as all commands that treats container are *root* commands (`run`, `create`, `start`, …).

We need inputs from @jhowardmsft, @swernli or other Microsoft guys for how and what could be supported for the Windows daemon (will it be possible to change memory constraint and such at *runtime* or not, …). 

To go forward :
- The command should be `set` (as this PR does).
- The allowed changes are passed as flags : e.g. `--memory=1Gb --cpushare=…` (as this PR does).
- There is one flag for each attribute of the `Resources` struct (and no more, no less).
- Depending on what is possible for windows, we'll have to think on how we implement that :stuck_out_tongue_closed_eyes:.

We also decided that **any future addition to the `set` command should come with real-life use case** that can't be done easily with orchestration tools or … We have to make it clear in this PR and in documentation that it is not open-bar :stuck_out_tongue_closed_eyes:.

Pinged people :wink: and @hqhq, WTDY?

thaJeztah commented:
@vdemeester thanks, I think that summarizes it properly; If we want to add additional "mutable" properties in future, we expect those properties to either *be* inside HostConfig, or *moved to* HostConfig, so that only "runtime" properties are mutable.

For now, we'll only support updating properties for a single container at a time, but the chosen syntax (`docker set <container> --option --option`) would potentially allow changing multiple container, but that would require a clear use-case

hqhq commented:
@vdemeester @thaJeztah Glad to see we got some progress here, I'm definitely +1 on the conclusion, I'll rebase this PR in the few days, and we can wait for windows guy's input in the mean time.

Thanks for your effort. :+1:

jhowardmsft commented:
All us Windows folks are out until 30th...

tiborvass commented:
One tiny addition is: we couldn't decide on setting labels therefore we left it out. Labels could be a special case in the sense that they are metadata. One thing that complicates things, is that it is part of the container Config, and not HostConfig. And like @thaJeztah said, we do not want to modify Config on runtime. Maybe HostConfig should be renamed to MutableConfig? Just throwing ideas.

thaJeztah commented:
@tiborvass yes, I was thinking of that earlier (labels) let's discuss that on https://github.com/docker/docker/issues/15496

> Maybe HostConfig should be renamed to MutableConfig? Just throwing ideas.

How about `RunTimeConfig` or `RunConfig` (also just throwing ideas)? That should probably be left for another PR though

duglin commented:
I think its implied, but just to be clear... the format of the various resources that can be set (e.g. `--memory`) will match the format/syntax of what's available on `docker run` and `docker create`, right? Just to ensure we have consistency.

tiborvass commented:
@duglin yes

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+func (cli *DockerCli) CmdSet(args ...string) error {
+	cmd := Cli.Subcmd("set", []string{"CONTAINER [CONTAINER...]"}, "Update properties of one or more containers", true)
```
Is there any way to reuse runconfig.Parse() so we don't have to define these flags in two places?

vdemeester replied:
That's a very good point :+1:

hqhq replied:
@duglin @vdemeester There was a try: https://github.com/docker/docker/pull/12235
Sadly it didn't workout, I'll try again when we get this settled. I think that can be done incrementally :)

hqhq commented:
@tiborvass @thaJeztah Isn't the conclusion we only `set` resources? I'm working on touch resources only since it's in a separate struct, is it still necessary that we handle `HostConfig`?
Anyway, I'll rebase it ASAP, have some dependency PR on going.

vdemeester commented:
@hqhq for the first version of the `set` command, and for this PR, only `Resource` struct. We'll discuss the labels in another PR I think (and how we handle them, whether we copy it or not at runtime, etc…).

hqhq commented:
@tiborvass @thaJeztah @duglin @vdemeester Rebased, PTAL.

thaJeztah commented:
Thanks @hqhq!

@vdemeester I forgot what we discussed on that, but did we decide if making changes via `docker set` would *persist*? i.e., would those changes be permanent (updated in the container's JSON) or be reverted when a container is restarted?

duglin commented:
I'd vote to persist them.  If I increased the memory of my container but then the container was restarted w/o my knowledge (perhaps the host was rebooted) then I'd be kind of surprised to not see the new memory size being used - and I'd have no indication that it was reverted.

thaJeztah commented:
@duglin yes, I think so; just checking with what was decided (because it wasn't mentioned in @vdemeester's comment)

duglin commented on daemon/container_unix.go:
```diff
@@ -1455,6 +1455,74 @@ func (container *Container) ipcMounts() []execdriver.Mount {
+	if resources.CPUPeriod != 0 {
+		cResources.CPUPeriod = resources.CPUPeriod
+	}
+	if resources.CPUQuota != 0 {
```
In general I'm not sure how we detect "0" (zero) from "not specified" in cases where zero is a valid value.

cpuguy83 replied:
Would need to use an int pointer instead of just int.

duglin replied:
yep that would work, but I think the default golang json parser is going to get in the way because I don't think, by default, it tells you if a particular property is in the json or not - all you see is zero in the struct afterwards. I think it might require a more specialized parser and then we'd need to make sure we only include "changed" fields in the json, not the entire struct.  Either that, or do a GET to get to current values and then we can pass them all, but the extra message exchange and risk of one of those "current values" changes on you is a problem.  I'd prefer a delta-update.

cpuguy83 replied:
No, we currently do this in other places where 0 is a valid value.
http://play.golang.org/p/Nbh5diB0ey

duglin replied:
ah, change the struct itself - sorry misread what you were saying.  nice!

jhowardmsft commented on integration-cli/docker_cli_set_test.go:
```diff
@@ -0,0 +1,59 @@
+package main
```
Can this be in a _unix file while set is not implemented on Windows?

hqhq replied:
Sure, will do that.

cpuguy83 commented:
Can we add tests using the API as well and hitting at least some of the weird cases, like MemorySwap?

Also needs a rebase :(

hqhq commented:
@jhowardmsft @cpuguy83 Updated and rebased.

duglin commented on container/container_unix.go:
```diff
@@ -550,6 +551,75 @@ func (container *Container) IpcMounts() []execdriver.Mount {
+
+	resources := hostConfig.Resources
+	cResources := &container.HostConfig.Resources
+	if resources.BlkioWeight != 0 {
```
What happened to the discussion of using pointers to ints so that we know if they specified anything at all since you can't tell if "0" means they want "0" or they didn't specify anything at all?

hqhq replied:
@duglin I think the strategy in Docker for now is to change all these fields whose default values have specific meanings to points, while leave the others.
Now since specs is trying to change all resource fields to points, maybe we can do that in Docker someday as well.

duglin replied:
I'm not sure what you mean by "points", can you elaborate?

hqhq replied:
Sorry, typo, I meant `pointers`, PR from specs: https://github.com/opencontainers/specs/pull/233

duglin replied:
ok but in this PR don't we need to change, at least some, of these to pointers so that we know whether the user really meant zero instead of "not specified" ?

hqhq replied:
All fields docker set touched in this PR don't have the problem (default value have special meaning). And change fields in `Resources` needs to touch other codes like start and create, I think that's out of scope. So I prefer we do that in a separate PR. WDYT?

duglin replied:
I might be misunderstanding but I don't think this PR can be merged w/o the use of pointers. Take as an example "resources.MemorySwap", I think that will be zero if not set on the "docker set" command line, but zero is a valid value according to the docs. Wouldn't this need a pointer to know if it was included in the json?

hqhq replied:
Zero is not valid value for `MemorySwap`, I think what you had in mind is actually `MemorySwappiness`, which is already a pointer. But it's not included in this PR. (It should, but new options are keep coming during this PR's development, so I didn't cache up all resource options, I hope it's OK we can add the rest of them in a separate PR)

duglin replied:
Ah ok thanks

duglin commented on integration-cli/docker_cli_set_unix_test.go:
```diff
@@ -0,0 +1,61 @@
+	file := "/sys/fs/cgroup/memory/memory.limit_in_bytes"
+	out, _ := dockerCmd(c, "exec", name, "cat", file)
+	c.Assert(strings.TrimSpace(out), checker.Equals, "524288000")
+}
```
I'd like to see a test where we set a value via set() and then check to make sure other values (that were not touched during the set() ) are not changed due to the "zero processing" I mentioned in a previous comment.

duglin commented:
overall I think it looks good - just a few comments around "zero processing".

vdemeester commented:
Looks good. We probably should add more tests, using `set` while running (and restarting it or not), while stopped or paused, etc… — maybe not one for each flag but *meh* I wouldn't be against it either.

hqhq commented:
@duglin @vdemeester Will add more tests, thanks for reviewing.

thaJeztah commented:
@hqhq thanks, let us know when they're added, hopefully we can then move to docs review soon

hqhq commented:
@duglin @vdemeester @thaJeztah Rebased and added more tests.

duglin commented on integration-cli/docker_cli_set_unix_test.go:
```diff
@@ -0,0 +1,121 @@
+
+	file := "/sys/fs/cgroup/cpuset/cpuset.cpus"
+	out, _ = dockerCmd(c, "exec", name, "cat", file)
+	c.Assert(strings.TrimSpace(out), checker.Equals, "0")
```
I think this test would be better if cpuset-cpus was set to "1" on line 98 and then we checked to make sure it was still "1" at this point.

hqhq replied:
I understand the concern, but the problem is it'll fail on machines which have only one cpu, so I changed it to set and check cpu shares, that'll fix the issue as well.

vdemeester commented:
> @vdemeester I forgot what we discussed on that, but did we decide if making changes via `docker set` would *persist*? i.e., would those changes be permanent (updated in the container's JSON) or be reverted when a container is restarted?

I actually don't remember (and it's not on my notes :sweat_smile:). But I would vote +1 as @duglin.

@hqhq tests looks good :wink:

unclejack commented:
@duglin @vdemeester We agreed that we would persist the changes and that it would be the operator's responsibility to change the settings back if they wish to switch to different settings.

hqhq commented:
Yeah, that's what this PR does :)

hqhq commented:
ping, any LGTMs if it looks good? Hope we can have this in 1.10 :pray: :sweat_smile:

duglin commented:
@hqhq sorry for the delay, will try to review it today. I did play with it yesterday and it seemed to work as expected, so its looking good so far....

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,100 @@
+	flKernelMemory := cmd.String([]string{"-kernel-memory"}, "", "Kernel memory limit")
+	cmd.Require(flag.Min, 1)
+
+	cmd.ParseFlags(args, true)
```
I do: `docker set 5c67` it will send a request to the server but nothing will happen, so this turns into a test to see if a container with that ID exists or not. I'm thinking we should generate an error if no flags are provided - we should be able to check this via: `cmd.NFlag() == 0` I think.  What do people think?

vdemeester replied:
I do agree with @duglin

duglin replied:
If we do add this check then let's also add a test to make for it.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,100 @@
+		}
+	}
+	if len(errNames) > 0 {
+		return fmt.Errorf("Error: failed to set containers: %v", errNames)
```
I'm pretty sure you added this to be consistent with other commands (and that's good) but we might want to consider another PR to remove this line - it seems redundant to me.

hqhq replied:
Yes I'm doing this to be consistent with other commands, not sure how to change, remove the whole `errNames`? Could you elaborate?

duglin replied:
I just don't think we need the entire if-statement. Up above, as we process each container, we print out an error or the container name - I don't see the point in repeating the list of containers that failed to be processed.  But let's save that for another PR.

hqhq replied:
Got it, I think it makes sense, will change it and open a new PR to change other commands, thanks.

duglin commented on integration-cli/docker_api_set_unix_test.go:
```diff
@@ -0,0 +1,43 @@
+// +build !windows
```
Don't we want the windows CLI to test against a linux server?

hqhq replied:
I did that before, checked with `testRequires(c, DaemonIsLinux)` for all cases, then @jhowardmsft suggested we move these tests to `_unix` test file, I think all `_unix` test files are built with !windows.
ping @jhowardmsft what would you prefer?

duglin replied:
@jhowardmsft are we testing with windows clients against linux servers?  I thought we were.

duglin commented on integration-cli/docker_cli_set_unix_test.go:
```diff
@@ -0,0 +1,121 @@
+// +build !windows
```
same here - I think we want this to run on windows CLI too - just with a linux server.

duglin commented on integration-cli/docker_cli_set_unix_test.go:
```diff
@@ -0,0 +1,121 @@
+	name := "test-set-container"
+	file := "/sys/fs/cgroup/memory/memory.limit_in_bytes"
+	dockerCmd(c, "run", "-d", "--name", name, "-m", "300M", "busybox", "cat", file)
+	dockerCmd(c, "wait", name)
```
No biggie but I think you could remove this "wait" call if you removed the "-d" from the "run" command.

duglin commented:
just a few comments - otherwise it looks good. 
Just waiting to see the responses before the formal "lgtm"

hqhq commented:
@duglin @vdemeester Apart from test for windows we need input from @jhowardmsft , updated other issues.

vdemeester commented:
I'm wondering something : setting kernel memory seems to only work when the container is stopped, has been started with `--kernel-memory` or has been restarted after setting `kernel-memory` (it will error out with a `Device or resource busy` otherwise) :

```bash
$ docker run -d nginx
669fc12b78fd897a679a6cf1aacb960db591ba32084991227b745cc51bcbfd4f
$ docker set --kernel-memory 700Mb 669
Failed to set resources of container (669): Error response from daemon: write /sys/fs/cgroup/memory/docker/669fc12b78fd897a679a6cf1aacb960db591ba32084991227b745cc51bcbfd4f/memory.kmem.limit_in_bytes: device or resource busy
$ docker stop 669
669
$ docker start 669
669
$ docker set --kernel-memory 700Mb 669
669
```

Even if it fails it is written into the `hostconfig.json` file so it feels almost ok but I'm not sure what to do about that.

hqhq commented:
@vdemeester I see, kernel memory has to be set with no processes in the cgroup, so we can't set `--kernel-memory` to a running container, but a stopped container would be fine. I'll update this, thanks for testing out!

hqhq commented:
FYI, it's kernel limit, see: https://github.com/torvalds/linux/blob/master/Documentation/cgroups/memory.txt#L284-L286

hqhq commented:
@vdemeester Updated
- added check for kernel memory set and running container
- added docs about kernel memory set
- added test case for kernel memory set

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,104 @@
+	cmd.Require(flag.Min, 1)
+	cmd.ParseFlags(args, true)
+	if cmd.NFlag() == 0 {
+		return fmt.Errorf("You should always set one or more flags when use this command.")
```
Minor thing but going back to a previous life where RFC2119 ruled everything, can you change this to something like:
`You must provide one or more flags when using this command` ?  The "should" doesn't mean its required, so "must" is more appropriate.

duglin commented on api/client/set.go:
```diff
@@ -0,0 +1,104 @@
+	var errs []string
+	for _, name := range names {
+		if err := cli.client.ContainerSet(name, hostConfig); err != nil {
+			errs = append(errs, fmt.Sprintf("Failed to set resources of container (%s): %s", name, err))
```
Like the other PR, I'd prefer to just print them in order - so add the "fmt.Fprintf" here and remove "errs".

duglin replied:
For now go ahead and make this like the other commands - I forgot that we need to return an error when things go wrong instead of just printing things immediately. Which means the order in which the IDs, or errors, will be printed won't match the cmd line args and that might be problematic for people.  So, I'd prefer to have that discussion as part of the other PR and for now make this one consistent with how the other cmds work.

duglin commented:
couple of minor suggestions, but other than that, LGTM!  thanks for the hard work!

hqhq commented:
@duglin Thanks, updated error message, for the other issue, I'll leave it to https://github.com/docker/docker/pull/18768 , I'll update when we have consensus there.

duglin commented on daemon/set.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	if container.RemovalInProgress || container.Dead {
+		return fmt.Errorf("Container is marked for removal and cannot be set.")
```
Very minor thing but can we wrap "set" in quotes?  W/o something, like quotes, it reads kind of funny - when someone first reads this they may ask "set to what?"  When really what we're trying to say is that the "set" function can not be used at this time.  Another option is to use the word "updated" but I think keeping it consistent with the action is probably better.

duglin commented:
couple of minor comments but I think we're ready, once @jhowardmsft gives his opinion on the testcase questions.

thaJeztah commented:
Failing tests should be solved by https://github.com/docker/docker/pull/18820, so if you can rebase this PR, those should be fixed

hqhq commented:
@duglin Updated.

ping @jhowardmsft for the test case question.

thaJeztah commented:
ping @vdemeester @duglin if you can review and "lgtm" (if it lgty), then feel free to move it to docs-review.

Given that @jhowardmsft is on vacation until the end of the year, we should move forward without his input. If needed, we can make changes in a follow up PR

vdemeester commented:
LGTM 🐲

duglin commented:
@hqhq go ahead and remove the
```
// +build !windows
``` 
and let's see if it still passes the windows build. If so then we've move forward with it.

icecrime commented:
Ping @moxiegirl as this has been waiting waaaaaaaaaaaay too long :-)

hqhq commented:
@duglin I tried remove that line, but seems we'll get a lot of error like 
`undefined: memoryLimitSupport` which is in the `testRequires`.

Since cgroup options in for `docker run` also tested under `// +build !windows`, I think it's OK we keep it consistent for now. And figure out "windows cli and linux server" scenario later, WDYT?

duglin commented:
@hqhq that's too bad those checks work that way. Oh well.
ok LGTM
Just needs a rebase.

moxiegirl commented:
@perhapszzy Thank you for the contribution.  Need to add this command to the `reference/commandline/index.md` file --- under Container commands maybe.

vdemeester commented:
@moxiegirl I think you meant @hqhq :stuck_out_tongue_closed_eyes: 

Needs a rebase too :sweat_smile: (hopefully the last one !)

moxiegirl commented:
@vdemeester I did..Sorry gang still looking at the command. Sec with that review. Wish I had gotten to the party earlier --- I kinda like `update` or `limit` for a command name.... Oh well.  Sec while I finish out this review.

moxiegirl commented:
I think we should update this diagram also: [Docker Events](https://docs.docker.com/engine/reference/api/docker_remote_api/#docker-events)  the source is in `reference/api/images` --- if you want, you can markup changes and we can carry the image change for you.

vdemeester commented:
@moxiegirl `limit` is a little bit *limiting* in the sense that it could be to do other thing that *limiting* stuff (I think of labels for example). `update` I like a lot though :wink: — it's never too late to get to the party :stuck_out_tongue_closed_eyes:.

thaJeztah commented:
Update sounds like a good candidate,

moxiegirl commented on docs/reference/commandline/set.md:
```diff
@@ -0,0 +1,30 @@
+<!--[metadata]>
```
@hqhq I think maybe more words. :-D How often do I say that?

---> https://gist.github.com/moxiegirl/cc09673bb8b0c1cfca99

<!--[metadata]>
+++
title = "set"
description = "The set command description and usage"
keywords = ["resources, set, dynamically"]
[menu.main]
parent = "smn_cli"
+++
<![end-metadata]-->

## set

    Usage: docker set [OPTIONS] CONTAINER [CONTAINER...]

    Sets container resource limits

      --help=false               Print usage
      --blkio-weight=0           Block IO weight (relative weight)
      --cpu-shares=0             CPU shares (relative weight)
      --cpu-period=0             Limit the CPU CFS (Completely Fair Scheduler) period
      --cpu-quota=0              Limit the CPU CFS (Completely Fair Scheduler) quota
      --cpuset-cpus=""           CPUs in which to allow execution (0-3, 0,1)
      --cpuset-mems=""           Memory nodes (MEMs) in which to allow execution (0-3, 0,1)
      -m, --memory=""            Memory limit
      --memory-reservation=""    Memory soft limit
      --memory-swap=""           Total memory (memory + swap), '-1' to disable swap
      --kernel-memory=""         Kernel memory limit: container must be stopped 

The `docker set` command dynamically sets container resources.  Use this command to prevent containers from consuming too many resources from their Docker host.  With a single command, you can place limits on a single container or on many. To specify more than one container, provide space-separated list of container names or IDs.

With the exception of the `--kernel-memory` value, you can specify these options on a running or a stopped container. You can only set  `--kernel-memory` on a stopped container. When you run `docker set` on stopped container, the next time you restart it, the container uses those values.

moxiegirl commented on man/docker-set.1.md:
```diff
@@ -0,0 +1,78 @@
+% DOCKER(1) Docker User Manuals
```
@hqhq Please reuse the text (*not* the metadata :-) from the `set.md` info above.

moxiegirl commented:
+1 for an informative error message if they use `--kernel-memory` on a running container. @vdemeester previous comment still holds I think. IMO, not having it forces people to read the docs

hqhq commented:
@moxiegirl Docs updated, thanks. I updated `event_state.gliffy` but can't export it to png, did you mean I can just update this file and you'll help export to png?

hqhq commented:
@moxiegirl @thaJeztah @vdemeester I like `set` better than `update` because I think `update` means you already have a value and you want to change it to something else. But for `docker set`, it's not only for properties you already used in `docker run`, it can be for a brand new property which is the first time you set it.

Anyway like I always said, I'm always open on the command names, but I think we've already fully discussed and `set` is the final winner? After all, it's a PR for more than 9 months. :wink:

vdemeester commented:
@hqhq we were thinking of `update` because it is **updating** the container — looking at the container and not at the value being set. This would allow to add some more flags to this same commands for stuff that could be added or removed (like *adding a label*, *removing a label*, etc…).

Looking at the other commands (`start`, `stop`, `pause`, `unpause`, `commit`, …) are all about "what we are doing to a container", and thus it feels like `update` would work on that matter.

I'm gonna quote @moxiegirl :wink: (others would be `update` in this case) :

> `set` has a "first time" implication that others don't.

The `set` command of this PR allows to set values that were or were not originally set. And almost all values set have default values (either `0` or empty string).

/cc @moxiegirl @thaJeztah @icecrime @duglin @ehazlett who were on the IRC discussion

hqhq commented:
@vdemeester I couldn't find the IRC discussion, so what's the conclusion there? Should I update the command to `update` then?

thaJeztah commented:
@hqhq IRC conversation can be found here; https://botbot.me/freenode/docker-maintainers/msg/56696314/. I think most people are in favor of renaming it to `docker update`

hqhq commented:
@thaJeztah Thanks for the link I'll update the command name. : )

thaJeztah commented:
Thank you @hqhq, sorry that the "name" discussion came so very late in the process, but we're still determined to get this merged ASAP

hqhq committed:
Implemet docker update command

It's used for updating properties of one or more containers, we only
support resource configs for now. It can be extended in the future.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@all Changed the command name to `docker update`

hqhq commented:
@moxiegirl We do have informative error message if people use `--kernel-memory` on a running container.

```
$ docker update --kernel-memory 50M b82062a7b587
Error response from daemon: Can not update kernel memory to a running container, please stop it first.
Error: failed to update resources of containers: [b82062a7b587]
```

vdemeester commented:
@hqhq Thanks !! :clap: LGTM again and again :wink:

thaJeztah commented:
docs LGTM! I think we're there, and it's ready to merge! :dancers: 

Thank you so much @hqhq, very happy to see this ready. Thanks for being so patient with us!

I restarted the Windows CI for good measure, because it timed out

thaJeztah commented:
sigh, flaky tests I think;

```
FAIL: docker_cli_events_test.go:415: DockerSuite.TestEventsStreaming

docker_cli_events_test.go:452:
    c.Fatal(observer.TimeoutError(containerID, "create"))
... Error: failed to observe event `create` for f0ad173280a3e918743be743bd661513b3bf9448d0eff85fd85f07e02a782528
2015-12-28T14:31:40.073268700Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) die
2015-12-28T14:31:40.276026700Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) stop
2015-12-28T14:31:40.513889100Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) destroy
```

and

```
FAIL: docker_cli_events_test.go:415: DockerSuite.TestEventsStreaming

docker_cli_events_test.go:452:
    c.Fatal(observer.TimeoutError(containerID, "create"))
... Error: failed to observe event `create` for f0ad173280a3e918743be743bd661513b3bf9448d0eff85fd85f07e02a782528
2015-12-28T14:31:40.073268700Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) die
2015-12-28T14:31:40.276026700Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) stop
2015-12-28T14:31:40.513889100Z 4a87a239201ddd2feb3843634fcd88435fb290f822f1d179458ee0093f6e9020: (from busybox) destroy
```

moxiegirl commented:
@hqhq Thank you for the work.

thaJeztah commented:
@moxiegirl I noticed some tiny nits in the docs, but will create a follow up PR to address those. I didn't think it was worth blocking the merge on those :)

moxiegirl commented:
@thaJeztah Thanks.  You notice I had not actually done an LGTM on this one...it was in a hurry I believe.

frakky commented:
Hey,
as this command will also handle changing ports on the fly in the future, how would I approach the implementation of that? 
I would like to help creating the port mapping feature but I don't really know where to start.

thaJeztah commented:
@frakky I think there's already an issue open for that feature; https://github.com/docker/docker/issues/2045. If you need help implementing, you can ask for assistance in the #docker-dev IRC channel. Keep in mind that it's currently very busy for the maintainers, because code-freeze for 1.10 is this week.
